src/main/java/org/gridsuite/voltageinit/server/computation/service/AbstractWorkerService.java

```diff
@@ -0,0 +1,241 @@
+                                    ReportService reportService,
+                                    T resultService,
+                                    ExecutionService executionService,
+                                    AbstractComputationObserver<S, P> observer,
+                                    ObjectMapper objectMapper) {
+        this.networkStoreService = networkStoreService;
+        this.notificationService = notificationService;
+        this.reportService = reportService;
+        this.resultService = resultService;
+        this.executionService = executionService;
+        this.observer = observer;
+        this.objectMapper = objectMapper;
+    }
+
+    protected PreloadingStrategy getNetworkPreloadingStrategy() {
+        return PreloadingStrategy.COLLECTION;
+    }
+
+    protected Network getNetwork(UUID networkUuid, String variantId) {
+        Network network;
+        try {
+            network = networkStoreService.getNetwork(networkUuid, getNetworkPreloadingStrategy());
+            String variant = StringUtils.isBlank(variantId) ? VariantManagerConstants.INITIAL_VARIANT_ID : variantId;
+            network.getVariantManager().setWorkingVariant(variant);
+        } catch (PowsyblException e) {
+            throw new ResponseStatusException(HttpStatus.NOT_FOUND, e.getMessage());
+        }
+        return network;
+    }
+
+    protected void cleanResultsAndPublishCancel(UUID resultUuid, String receiver) {
+        resultService.delete(resultUuid);
+        notificationService.publishStop(resultUuid, receiver, getComputationType());
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info("{} (resultUuid='{}')",
+                    NotificationService.getCancelMessage(getComputationType()),
+                    resultUuid);
+        }
+    }
+
+    private void cancelAsync(CancelContext cancelContext) {
+        lockRunAndCancel.lock();
+        try {
+            cancelComputationRequests.put(cancelContext.getResultUuid(), cancelContext);
+
+            // find the completableFuture associated with result uuid
+            CompletableFuture<S> future = futures.get(cancelContext.getResultUuid());
+            if (future != null) {
+                future.cancel(true);  // cancel computation in progress
+            }
+            cleanResultsAndPublishCancel(cancelContext.getResultUuid(), cancelContext.getReceiver());
+        } finally {
+            lockRunAndCancel.unlock();
+        }
+    }
+
+    protected abstract AbstractResultContext<R> fromMessage(Message<String> message);
+
+    protected boolean resultCanBeSaved(S result) {
+        return result != null;
+    }
+
+    public Consumer<Message<String>> consumeRun() {
+        return message -> {
+            AbstractResultContext<R> resultContext = fromMessage(message);
+            try {
+                AtomicReference<Long> startTime = new AtomicReference<>();
+                startTime.set(System.nanoTime());
```

Comment:
No need of an atomic reference here, we stay in same thread.
```suggestion
                long startTime = System.nanoTime();
```